verification/src/compiler/version.rs

```diff
@@ -123,6 +151,22 @@ impl Display for CompilerVersion {
     }
 }
 
+impl Ord for CompilerVersion {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (self.version(), self.date(), self.commit()).cmp(&(
```

Comment:
How we compare `date` with release and nightly? For example, `v0.4.0` and `v0.4.0-nightly..`? 
Looking at docs, it seems, like `None` will be less, than `Some()`, so we'll have
`0.3.9, 0.4.0, 0.4.0-nightly, 0.4.1,...`, which isn't right

And actually I'd really prefer not to use general methods here and compare variants one by one, something like
```rust

// impl Ord for ReleaseVersion and NightlyVersion

// release
(&left.version, &left.commit).cmp(&(&right.version, &right.commit))

// nightly
(&left.version, &left.date, &left.commit).cmp(...)

// for CompilerVersion
fn cmp_release_nightly(left: Release, right: Nightly) -> Ordering {
  (&left.version, 1/*, &left.commit*/).cmp(&(&right.version, 0/*, &right.commit*/)) // any release is later than any nightly with the same version
  // you can leave out the commit
}

fn cmp(&self, other: &Self) -> Ordering {
  match self {
    ReleaseVersion(left) => {
      match other {
        ReleaseVersion(right) => {
          left.cmp(right)
        },
        NightlyVersion(right) => {
          cmp_release_nightly(left, right)
        },
      }
    },
    NightlyVersion(left) => {
      match other {
        ReleaseVersion(_) => {
          cmp_release_nightly(right, left).reverse()
        }
        NightlyVersion(right) => {
          left.cmp(right)
        }
      }
    }
  }
}
```

Reply:
Or, if you want your way, you can reorder fields in our structs and enum, and/or make some new structs, if needed, for which `derive(Ord)` will do the sorting
Something similar to your impl, but more cleaner
```rust
#[derive(PartialOrd, Ord)]
/*private*/ struct VersionSort {
  version: &Version,
  release: bool,
  date: Option<&Date>,
  commit: &[u8],
}

// and then
impl From<ReleaseVersion> for VersionSort;
impl From<NightlyVersion> for VersionSort;
impl From<CompilerVersion> for VersionSort;

fn cmp(self, other) -> Ordering {
  self.into::<VersionSort>.cmp(other.into())
}
```

Reply:
@rimrakhimov what do you think?

Reply:
@sevenzing proposed the idea to introduce a new function `CompilerVersion::is_release(&self) -> bool` and use that value as a second argument of a tuple. Personally, I support that idea, as it seems to be clearer.